File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int>ans(n,-1);
        stack<int>stk; 
        for(int i=2*n - 1 ; i >= 0 ; i--){
            while(!stk.empty() && nums[stk.top()] <= nums[i%n]){
                stk.pop();
            }
            if(!stk.empty())ans[i%n]=nums[stk.top()];
            stk.push(i%n);
        }
        return ans; 
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;
        int[] ans=new int[n];
        Stack<Integer>s=new Stack<Integer>();
        for(int i=2*n - 1 ; i>=0 ; i--){
            while(!s.isEmpty() && nums[s.peek()%n]<=nums[i%n]){
                s.pop();
            }
            ans[i%n]=(s.isEmpty()) ? -1 : nums[s.peek()%n];
            s.push(i%n);
        }
        return ans;
    }
}


// 1 2 1 4 5 9 

// 0 1 2 3 4  5 
// 6 7 8 9 10 11 
 

